File: CfxGenerator/ApiTypes/WrapperOnly/CefStructPtrArrayType.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



''' <summary>
''' Array of struct pointers, cef_struct **array
''' </summary>
''' <remarks></remarks>
Public Class CefStructPtrArrayType
    Inherits CefStructPtrPtrType

    Public ReadOnly StructArg As Argument
    Public ReadOnly CountArg As Argument

    Sub New(structArg As Argument, countArg As Argument)
        MyBase.New(structArg.ArgumentType.AsCefStructPtrPtrType)
        Me.StructArg = structArg
        Me.CountArg = countArg
    End Sub

    Public Overrides ReadOnly Property PublicSymbol As String
        Get
            Return Struct.ClassName & "[]"
        End Get
    End Property

    Public Overrides ReadOnly Property ProxySymbol As String
        Get
            Return "ulong[]"
        End Get
    End Property

    Public Overrides ReadOnly Property RemoteSymbol As String
        Get
            Return Struct.RemoteClassName & "[]"
        End Get
    End Property

    Public Overrides ReadOnly Property PublicUnwrapExpression(var As String) As String
        Get
            If StructArg.Index < CountArg.Index Then
                Return String.Format("{0}_pinned.PinnedPtr, {0}_length", var)
            Else
                Return String.Format("{0}_length, {0}_pinned.PinnedPtr", var)
            End If
        End Get
    End Property

    Public Overrides ReadOnly Property ProxyWrapExpression(var As String) As String
        Get
            Return String.Format("CfxArray.GetProxyIds<{0}>({1})", Struct.ClassName, var)
        End Get
    End Property

    Public Overrides ReadOnly Property ProxyUnwrapExpression(var As String) As String
        Get
            Return var & "_unwrapped"
        End Get
    End Property

    Public Overrides ReadOnly Property RemoteWrapExpression(var As String) As String
        Get
            Return String.Format("CfxArray.GetCfrObjects<{0}>({1}, remoteRuntime, {0}.Wrap)", Struct.RemoteClassName, var)
        End Get
    End Property

    Public Overrides ReadOnly Property RemoteUnwrapExpression(var As String) As String
        Get
            Return var & "_unwrapped"
        End Get
    End Property


    Public Overrides ReadOnly Property PublicEventConstructorSignature(var As String) As String
        Get
            Return StructPtr.PublicEventConstructorSignature(var) & ", " & CountArg.PublicEventConstructorSignature
        End Get
    End Property

    Public Overrides ReadOnly Property PublicEventConstructorCall(var As String) As String
        Get
            Return StructPtr.PublicEventConstructorCall(var) & ", " & CountArg.PublicEventConstructorCall
        End Get
    End Property

    Public Overrides Sub EmitPrePublicCallStatements(b As CodeBuilder, var As String)

        b.AppendLine("int {0}_length;", var)
        b.AppendLine("IntPtr[] {0}_ptrs;", var)
        b.BeginIf("{0} != null", var)
        b.AppendLine("{0}_length = {0}.Length;", var)
        b.AppendLine("{0}_ptrs = new IntPtr[{0}_length];", var)
        b.BeginBlock("for(int i = 0; i < {0}_length; ++i)", var)
        b.AppendLine("{0}_ptrs[i] = {1}.Unwrap({0}[i]);", var, Struct.ClassName)
        b.EndBlock()
        b.BeginElse()
        b.AppendLine("{0}_length = 0;", var)
        b.AppendLine("{0}_ptrs = null;", var)
        b.EndBlock()
        b.AppendLine("PinnedObject {0}_pinned = new PinnedObject({0}_ptrs);", var)
    End Sub

    Public Overrides Sub EmitPostPublicCallStatements(b As CodeBuilder, var As String)
        b.AppendLine("{0}_pinned.Free();", var)
    End Sub

    Public Overrides Sub EmitPublicEventArgFields(b As CodeBuilder, var As String)
        b.AppendLine("internal IntPtr[] m_{0};", var)
        b.AppendLine("internal {0}[] m_{1}_managed;", Struct.ClassName, var)
    End Sub

    Public Overrides Sub EmitPublicEventCtorStatements(b As CodeBuilder, var As String)
        b.AppendLine("m_{0} = new IntPtr[{1}];", var, CountArg.VarName)
        b.BeginIf("{0} > 0", CountArg.VarName)
        b.AppendLine("System.Runtime.InteropServices.Marshal.Copy({0}, m_{0}, 0, {1});", var, CountArg.VarName)
        b.EndBlock()
    End Sub

    Public Overrides Sub EmitPublicEventArgGetterStatements(b As CodeBuilder, var As String)
        b.BeginIf("m_{0}_managed == null", var)
        b.AppendLine("m_{0}_managed = new {1}[m_{0}.Length];", var, Struct.ClassName)
        b.BeginBlock("for(int i = 0; i < m_{0}.Length; ++i)", var)
        b.AppendLine("m_{0}_managed[i] = {1}.Wrap(m_{0}[i]);", var, Struct.ClassName)
        b.EndBlock()
        b.EndBlock()
        b.AppendLine("return m_{0}_managed;", var)
    End Sub

    Public Overrides Sub EmitPostPublicRaiseEventStatements(b As CodeBuilder, var As String)
        b.BeginIf("e.m_{0}_managed == null", var)
        b.BeginBlock("for(int i = 0; i < {1}; ++i)", var, CountArg.VarName)
        b.AppendLine("CfxApi.cfx_release(e.m_{0}[i]);", var)
        b.EndBlock()
        b.EndBlock()
    End Sub

    Public Overrides Sub EmitPreProxyCallStatements(b As CodeBuilder, var As String)
        b.AppendLine("{0}[] {1}_unwrapped;", Struct.ClassName, var)
        b.BeginIf("{0} != null", var)
        b.AppendLine("{0}_unwrapped = new {1}[{0}.Length];", var, Struct.ClassName)
        b.BeginBlock("for(int i = 0; i < {0}.Length; ++i)", var)
        b.AppendLine("{0}_unwrapped[i] = {1};", var, StructPtr.ProxyUnwrapExpression(var & "[i]"))
        b.EndBlock()
        b.BeginElse()
        b.AppendLine("{0}_unwrapped = null;", var)
        b.EndBlock()
    End Sub

    Public Overrides Sub EmitPreRemoteCallStatements(b As CodeBuilder, var As String)
        b.BeginIf("{0} != null", var)
        b.AppendLine("call.{0} = new ulong[{0}.Length];", var)
        b.BeginBlock("for(int i = 0; i < {0}.Length; ++i)", var)
        b.AppendLine("call.{0}[i] = {1};", var, StructPtr.RemoteUnwrapExpression(var & "[i]"))
        b.EndBlock()
        b.EndBlock()
    End Sub

End Class

File: CfxGenerator/ApiTypes/WrapperOnly/CefStructPtrArrayType.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System;
using System.Diagnostics;

/// <summary>
/// Array of struct pointers, cef_struct **array
/// </summary>
/// <remarks></remarks>
public class CefStructPtrArrayType : CefStructPtrPtrType {
    public readonly Parameter StructArg;

    public readonly Parameter CountArg;

    public CefStructPtrArrayType(Parameter structArg, Parameter countArg)
        : base(structArg.ParameterType.AsCefStructPtrPtrType) {
        this.StructArg = structArg;
        this.CountArg = countArg;
    }

    public override string PublicSymbol {
        get { return Struct.ClassName + "[]"; }
    }

    public override string ProxySymbol {
        get { return "IntPtr[]"; }
    }

    public override string RemoteSymbol {
        get { return Struct.RemoteClassName + "[]"; }
    }

    public override string PublicUnwrapExpression(string var) {
        if(StructArg.Index < CountArg.Index) {
            return string.Format("{0}_pinned.PinnedPtr, {0}_length", var);
        } else {
            return string.Format("{0}_length, {0}_pinned.PinnedPtr", var);
        }
    }

    public override string ProxyWrapExpression(string var) {
        return string.Format("CfxArray.GetProxyIds<{0}>({1})", Struct.ClassName, var);
    }

    public override string ProxyUnwrapExpression(string var) {
        return var + "_unwrapped";
    }

    public override string RemoteWrapExpression(string var) {
        return string.Format("CfxArray.GetCfrObjects<{0}>(CfxRemoteCallContext.CurrentContext.connection, {1}, {0}.Wrap)", Struct.RemoteClassName, var);
    }

    public override string RemoteUnwrapExpression(string var) {
        return var + "_unwrapped";
    }

    public override string PInvokeCallbackParameter(string var) {
        if(Struct.IsRefCounted)
            return base.PInvokeCallbackParameter(var) + ", out int " + var + "_release";
        else
            return base.PInvokeCallbackParameter(var);
    }

    public override string NativeCallbackParameter(string var, bool isConst) {
        if(Struct.IsRefCounted)
            return base.NativeCallbackParameter(var, isConst) + ", int *" + var + "_release";
        else
            return base.NativeCallbackParameter(var, isConst);
    }

    public override string NativeCallbackArgument(string var) {
        if(Struct.IsRefCounted)
            return base.NativeCallbackArgument(var) + ", &" + var + "_release";
        else
            return base.NativeCallbackArgument(var);
    }

    public override void EmitPreNativeCallbackStatements(CodeBuilder b, string var) {
        if(Struct.IsRefCounted)
            b.AppendLine("int " + var + "_release;");
    }

    public override void EmitPostNativeCallbackStatements(CodeBuilder b, string var) {
        if(Struct.IsRefCounted) {
            b.BeginBlock("if({0}_release)", var);
            b.BeginBlock("for(size_t i = 0; i < {0}; ++i)", CountArg.VarName);
            b.AppendLine("{0}[i]->base.release((cef_base_ref_counted_t*){0}[i]);", var);
            b.EndBlock();
            b.EndBlock();
        }
    }

    public override void EmitPreNativeCallStatements(CodeBuilder b, string var) {
        Debug.Assert(Struct.IsRefCounted);
        b.BeginIf("{0}", CountArg.VarName);
        b.BeginBlock("for(size_t i = 0; i < {0}; ++i)", CountArg.VarName);
        b.AppendLine("if({0}[i]) ((cef_base_ref_counted_t*){0}[i])->add_ref((cef_base_ref_counted_t*){0}[i]);", StructArg.VarName);
        b.EndBlock();
        b.EndBlock();
    }

    public override void EmitPrePublicCallStatements(CodeBuilder b, string var) {
        b.AppendLine("UIntPtr {0}_length;", var);
        b.AppendLine("IntPtr[] {0}_ptrs;", var);
        b.BeginIf("{0} != null", var);
        b.AppendLine("{0}_length = (UIntPtr){0}.Length;", var);
        b.AppendLine("{0}_ptrs = new IntPtr[{0}.Length];", var);
        b.BeginBlock("for(int i = 0; i < {0}.Length; ++i)", var);
        b.AppendLine("{0}_ptrs[i] = {1}.Unwrap({0}[i]);", var, Struct.ClassName);
        b.EndBlock();
        b.BeginElse();
        b.AppendLine("{0}_length = UIntPtr.Zero;", var);
        b.AppendLine("{0}_ptrs = null;", var);
        b.EndBlock();
        b.AppendLine("PinnedObject {0}_pinned = new PinnedObject({0}_ptrs);", var);
    }

    public override void EmitPostPublicCallStatements(CodeBuilder b, string var) {
        b.AppendLine("{0}_pinned.Free();", var);
    }

    public override void EmitPublicEventArgFields(CodeBuilder b, string var) {
        b.AppendLine("internal IntPtr[] m_{0};", var);
        b.AppendLine("internal {0}[] m_{1}_managed;", Struct.ClassName, var);
    }

    public override void EmitRemoteEventArgFields(CodeBuilder b, string var) {
        b.AppendLine("internal {0}[] m_{1}_managed;", Struct.RemoteClassName, var);
    }

    public override void EmitPublicEventCtorStatements(CodeBuilder b, string var) {
        b.AppendLine("e.m_{0} = new IntPtr[(ulong){1}];", var, CountArg.VarName);
        b.BeginIf("e.m_{0}.Length > 0", var);
        b.AppendLine("System.Runtime.InteropServices.Marshal.Copy({0}, e.m_{0}, 0, (int){1});", var, CountArg.VarName);
        b.EndBlock();
    }

    public override void EmitPublicEventArgGetterStatements(CodeBuilder b, string var) {
        b.BeginIf("m_{0}_managed == null", var);
        b.AppendLine("m_{0}_managed = new {1}[m_{0}.Length];", var, Struct.ClassName);
        b.BeginBlock("for(int i = 0; i < m_{0}.Length; ++i)", var);
        b.AppendLine("m_{0}_managed[i] = {1}.Wrap(m_{0}[i]);", var, Struct.ClassName);
        b.EndBlock();
        b.EndBlock();
        b.AppendLine("return m_{0}_managed;", var);
    }

    public override void EmitRemoteEventArgGetterStatements(CodeBuilder b, string var) {
        b.BeginIf("m_{0}_managed == null", var);
        b.AppendLine("var {0} = new RemotePtr[(ulong)call.{1}];", var, CountArg.VarName);
        b.AppendLine("m_{0}_managed = new {1}[{0}.Length];", var, Struct.RemoteClassName);
        b.BeginIf("{0}.Length > 0", var);
        b.AppendLine("CfrRuntime.Marshal.Copy(new RemotePtr(connection, call.{0}), {0}, 0, {0}.Length);", var);
        b.BeginBlock("for(int i = 0; i < {0}.Length; ++i)", var);
        b.AppendLine("m_{0}_managed[i] = {1}.Wrap({0}[i]);", var, Struct.RemoteClassName);
        b.EndBlock();
        b.EndBlock();
        b.EndBlock();
        b.AppendLine("return m_{0}_managed;", var);
    }

    public override void EmitPostPublicRaiseEventStatements(CodeBuilder b, string var) {
        if(Struct.IsRefCounted)
            b.AppendLine("{0}_release = e.m_{0}_managed == null? 1 : 0;", var);
        else if(Struct.IsScoped) {
            // never reached, but if this is reached in the future, 
            // then emit code to dispose all array elements
            throw new Exception();
        }
    }

    public override void EmitPostRemoteRaiseEventStatements(CodeBuilder b, string var) {
        if(Struct.IsRefCounted)
            b.AppendLine("{0}_release = e.m_{0}_managed == null? 1 : 0;", var);
        else if(Struct.IsScoped) {
            // never reached, but if this is reached in the future, 
            // then emit code to dispose all array elements
            throw new Exception();
        }
    }

    public override void EmitSetCallbackArgumentToDefaultStatements(CodeBuilder b, string var) {
        if(Struct.IsRefCounted)
            b.AppendLine("{0}_release = 1;", var);
    }
    
    public override void EmitPreProxyCallStatements(CodeBuilder b, string var) {
        b.AppendLine("PinnedObject {0}_pinned = new PinnedObject({0});", var);
        b.AppendLine("var {0}_length = {0} == null ? UIntPtr.Zero : (UIntPtr){0}.LongLength;", var);
    }

    public override void EmitPreRemoteCallStatements(CodeBuilder b, string var) {
        b.BeginIf("{0} != null", var);
        b.AppendLine("call.{0} = new IntPtr[{0}.Length];", var);
        b.BeginBlock("for(int i = 0; i < {0}.Length; ++i)", var);
        b.AppendLine("if(!CheckConnection({0}[i], connection)) throw new ArgumentException(\"Render process connection mismatch.\", \"{1}[\" + i + \"]\");", CSharp.Escape(var), var);
        b.AppendLine("call.{0}[i] = {1};", var, StructPtr.RemoteUnwrapExpression(var + "[i]"));
        b.EndBlock();
        b.EndBlock();
    }
}